### PR TITLE
[UNR-885] FUniqueNetId support, KCP integration

### PR DESCRIPTION
**Pull request by improbable-valentyn**
**Contributions**: We are not currently taking public contributions - see our [contributions](CONTRIBUTING.md) policy. However, we are accepting issues and we do want your [feedback](../README.md#give-us-feedback).

-------

#### Description
Mostly David's work from [this PR](https://github.com/spatialos/UnrealGDK/pull/564), excluding the Locator v2 flow, which was extracted separately (here)[https://github.com/spatialos/UnrealGDK/pull/615].
#### Tests
Tested locally with StarterProject, ShooterGame.
#### Primary reviewers
@davedissian @m-samiec

### CODE REVIEW

File: SpatialGDK/Extras/schema/spawner.schema [hunk 1]
```diff
@@ -3,6 +3,8 @@ package unreal;
 
 type SpawnPlayerRequest {
     string url = 1;
+    bytes unique_id = 2;
```
Comment (Vatyx) [inline SpatialGDK/Extras/schema/spawner.schema]: Why bytes instead of string?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp [hunk 2]
```diff
@@ -27,12 +29,24 @@ void USpatialPlayerSpawner::Init(USpatialNetDriver* InNetDriver, FTimerManager*
 	NumberOfAttempts = 0;
 }
 
-void USpatialPlayerSpawner::ReceivePlayerSpawnRequest(FString URLString, const char* CallerAttribute, Worker_RequestId RequestId )
+void USpatialPlayerSpawner::ReceivePlayerSpawnRequest(Schema_Object* Payload, const char* CallerAttribute, Worker_RequestId RequestId )
 {
+	// Extract spawn parameters.
+	FString URLString = GetStringFromSchema(Payload, 1);
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Hmm, I preferred this function not having deserialization code. Could we have a type with the 3 parameters which are passed in?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp [hunk 3]
```diff
@@ -27,12 +29,24 @@ void USpatialPlayerSpawner::Init(USpatialNetDriver* InNetDriver, FTimerManager*
...
+
+	FName OnlinePlatformName = FName(*GetStringFromSchema(Payload, 3));
+
+	// Accept new player.
 	URLString.Append(TEXT("?workerAttribute=")).Append(UTF8_TO_TCHAR(CallerAttribute));
 
-	NetDriver->AcceptNewPlayer(FURL(nullptr, *URLString, TRAVEL_Absolute), false);
+	NetDriver->AcceptNewPlayer(FURL(nullptr, *URLString, TRAVEL_Absolute), UniqueId, OnlinePlatformName, false);
 
+	// Send response.
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Useless comment?

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h [hunk 4]
```diff
@@ -14,7 +14,7 @@ struct FConnectionConfig
 	FConnectionConfig()
 		: UseExternalIp(false)
 		, EnableProtocolLoggingAtStartup(false)
-		, LinkProtocol(WORKER_NETWORK_CONNECTION_TYPE_RAKNET)
+		, LinkProtocol(WORKER_NETWORK_CONNECTION_TYPE_KCP)
```
Comment (Vatyx) [inline SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h]: We want to default to KCP? That's intentional?

File: SpatialGDK/Extras/schema/spawner.schema (hunk 1 above)
Comment (improbable-valentyn) [reply SpatialGDK/Extras/schema/spawner.schema]: This is serialized data. `string` implies a valid Unicode string.

Comment (Vatyx) [reply SpatialGDK/Extras/schema/spawner.schema]: > String and bytes fields are both treated as bytes in the C API, because they are both treated in the same way over the wire. In SpatialOS, the only difference is how code is generated for them in the C++, C#, and Java SDK, and how they’re visualized in the inspector.

https://docs.improbable.io/reference/13.5/capi/serialization#strings-and-bytes

I think it should be string due to it usually being human readable.

File: SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h (hunk 4 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Public/Interop/Connection/ConnectionConfig.h]: According to David, KCP is better than RakNet in every way, so yes.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 5]
```diff
@@ -328,7 +328,8 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 			FString URLString = FURL().ToString();
 			URLString += TEXT("?workerAttribute=") + UnrealMetadataComponent->OwnerWorkerAttribute;
 
-			Connection = NetDriver->AcceptNewPlayer(FURL(nullptr, *URLString, TRAVEL_Absolute), true);
+			// TODO: Once we can checkout PlayerController and PlayerState atomically, we can grab the UniqueId and online subsystem type from PlayerState.
+			Connection = NetDriver->AcceptNewPlayer(FURL(nullptr, *URLString, TRAVEL_Absolute), FUniqueNetIdRepl(), FName(), true);
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: We could potentially put the UniqueId and PlatformName inside a handwritten schema component and attach that to the PlayerController entity.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp [hunk 6]
```diff
@@ -15,7 +15,9 @@ FSpatialNetBitReader::FSpatialNetBitReader(USpatialPackageMapClient* InPackageMa
 
 void FSpatialNetBitReader::DeserializeObjectRef(FUnrealObjectRef& ObjectRef)
 {
-	*this << ObjectRef.Entity;
+	int64 EntityId;
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp]: Why was this necessary?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp [hunk 7]
```diff
@@ -27,11 +29,22 @@ void USpatialPlayerSpawner::Init(USpatialNetDriver* InNetDriver, FTimerManager*
 	NumberOfAttempts = 0;
 }
 
-void USpatialPlayerSpawner::ReceivePlayerSpawnRequest(FString URLString, const char* CallerAttribute, Worker_RequestId RequestId )
+void USpatialPlayerSpawner::ReceivePlayerSpawnRequest(Schema_Object* Payload, const char* CallerAttribute, Worker_RequestId RequestId )
 {
+	// Extract spawn parameters.
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Worth adding a comment to which component this is referring to?

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp [hunk 8]
```diff
@@ -112,3 +135,21 @@ void USpatialPlayerSpawner::ReceivePlayerSpawnResponse(Worker_CommandResponseOp&
...
 	}
 }
+
+void USpatialPlayerSpawner::ObtainPlayerId(FUniqueNetIdRepl& OutUniqueId, FName& OutOnlinePlatformName)
+{
+	const FWorldContext* const WorldContext = GEngine->GetWorldContextFromWorld(NetDriver->GetWorld());
+	check(WorldContext->OwningGameInstance);
+
+	// This code is adapted from PendingNetGame.cpp:242
+	ULocalPlayer* LocalPlayer = WorldContext->OwningGameInstance->GetFirstGamePlayer();
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Move * into if statement.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Tracking Jira please :)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp [hunk 9]
```diff
@@ -779,39 +779,18 @@ void USpatialSender::ResolveOutgoingRPCs(UObject* Object)
 
 FString USpatialSender::GetOwnerWorkerAttribute(AActor* Actor)
 {
+	// This should only be executed on the server.
+	check(NetDriver->GetServerConnection() == nullptr);
```
Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialSender.cpp]:
```suggestion
	check(NetDriver->IsServer());
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp (hunk 6 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp]: `int64` resolves to a different type than `int64_t` (one was `long`, the other `long long`, don't remember which is which though)

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp (hunk 7 above)
Comment (improbable-valentyn) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialPlayerSpawner.cpp]: Not a component. It's the `PlayerSpawnRequest`, so I figured the function name `ReceivePlayerSpawnRequest` should be self-documenting enough?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp (hunk 6 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetBitReader.cpp]: Oh c++

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -328,7 +328,8 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 			FString URLString = FURL().ToString();
 			URLString += TEXT("?workerAttribute=") + UnrealMetadataComponent->OwnerWorkerAttribute;
 
-			Connection = NetDriver->AcceptNewPlayer(FURL(nullptr, *URLString, TRAVEL_Absolute), true);
+			// TODO: Once we can checkout PlayerController and PlayerState atomically, we can grab the UniqueId and online subsystem type from PlayerState.
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
			// TODO: Once we can checkout PlayerController and PlayerState atomically, we can grab the UniqueId and online subsystem type from PlayerState. UNR-933
```
